Implement CYW4373E USB device driver

julianuziemblo opened the pull request:
Driver for vendor-specific usbwlan devices, only available for Sterling LWB5+ USB Dongle (CYW4373E) for now.

JIRA: RTOS-1122

<!--- Provide a general summary of your changes in the Title above -->

## Description

## Motivation and Context

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `armv7m7-imxrt117x-evk`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work:
    - https://github.com/phoenix-rtos/phoenix-rtos-usb/pull/38
    - https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/750
    - https://github.com/phoenix-rtos/libphoenix/pull/466
- [x] I will merge this PR by myself when appropriate - after ^above fix is merged

julianuziemblo committed:
add usbwlan driver

Driver for vendor-specific usbwlan devices
Only available for Sterling LWB5+ USB Dongle for now.

JIRA: RTOS-1122

julianuziemblo commented on net/usbwlan.c:
```diff
@@ -0,0 +1,531 @@
+{
+	rbnode_t *node, *next;
+	usbwlan_dev_t *dev;
+
```
`_usbwlan_put()` returns the refcount of the node, so `usbwlan_free()` is only called when the refcount is 0. Additionally, `usbwlan_put()` cannot be used, because we're in the context of a locked mutex.

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,521 @@
+#include <usbhost.h>
+
+#include "sys/msg.h"
+#include "usb.h"
```
```suggestion
#include <sys/msg.h>
#include <usb.h>
```

nit but maybe move inclusion of `sys/msg.h` to the highest include block

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,521 @@
+	idtree_t devices;
+	unsigned msgport;
+	handle_t lock;
+	char msgthread_stacks[USBWLAN_N_MSGTHREADS][1024] __attribute__((aligned(8)));
```
Use camel case. You may as well call it `stacks` as it's rather unambiguous

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,521 @@
+
+	dev = calloc(1, sizeof(*dev));
+	if (dev == NULL) {
+		fprintf(stderr, "usbwlan: Not enough memory for new device\n");
```
I personally prefer to introduce `log_{info,warning,error,debug}` macros in new code as this makes the strings shorter and makes changes to the logging provider easier, but this is up to you

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,521 @@
+				usbwlan_put(dev);
+				msgRespond(usbwlan_common.msgport, &msg, rid);
+				endthread();
+				return;
```
`endthread()` should not return. Maybe replace `return` with `__builtin_unreachable()` to signify that instead

(we should mark `endthread` as `noreturn` at last)

julianuziemblo commented on net/usbwlan.h:
```diff
@@ -0,0 +1,40 @@
+		usbwlan_ctrl_in,
+		usbwlan_ctrl_out,
+		usbwlan_reg_read,
+		usbwlan_reg_write } type;
```
okay, changing to `/*clang-format off|on */` for now

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,542 @@
+		.destroy = usbwlan_destroy,
+	},
+	.filters = filters,
+	.nfilters = (sizeof(filters) / sizeof(filters[0])),
```
nit: unnecessary parenthesis

adamgreloch reviewed (commented):
Minor nitpicks, otherwise LGTM

julianuziemblo commented on net/usbwlan.c:
```diff
@@ -0,0 +1,542 @@
+	int ret;
+
+	usbwlan_common.msgport = -1;
+	usbwlan_common.lock = -1;
```
I did them for `usbwlan_destroy` to release only resources that were allocated, as it's called regardless of what fails.

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,721 @@
+
+	do {
+		if (completion->err != 0) {
+			ret = -abs(completion->err);
```
There is no invariant on the sign of `completion->err`?

julianuziemblo committed:
net/usbwlan: make read async so it can be interrupted

The user may want to close the device when a thread is in the middle of
an ongoing read. This can be achieved through added usbwlan_abort devctl.

JIRA: RTOS-1122

adamgreloch commented on net/usbwlan.c:
```diff
@@ -0,0 +1,658 @@
+}
+
+
+/* stub */
```
not anymore :p

julianuziemblo commented on net/usbwlan.c:
```diff
@@ -0,0 +1,658 @@
+		}
+
+		dev = usbwlan_get(msg.oid.id);
+		if (dev == NULL) {
```
why should I put when `usbwlan_get` fails to find device?

adamgreloch replied:
ah, right

julianuziemblo commented on net/usbwlan.c:
```diff
@@ -0,0 +1,658 @@
+			case mtDestroy:
+				usbwlan_put(dev);
+				msgRespond(usbwlan_common.msgport, &msg, rid);
+				endthread();
```
It's not that simple as there can be multiple msgthreads. I assume here that `mtDestroy` is only broadcast from `usbwlan_destroy`.

adamgreloch replied:
Then best to put this assumption in a comment here

ziemleszcz commented on net/usbwlan.h:
```diff
@@ -0,0 +1,38 @@
+#define _USBWLAN_H_
+
+#include <stdint.h>
+#include <unistd.h>
```
Why is `unistd.h` needed here?

julianuziemblo replied:
probably wanted `stddef.h` for `size_t`. will change

ziemleszcz commented on net/usbwlan.c:
```diff
@@ -0,0 +1,656 @@
+
+static int usbwlan_destroy(usb_driver_t *drv)
+{
+	for (size_t i = 0; i < USBWLAN_N_MSGTHREADS; i++) {
```
Please implement full cleanup of all resources or add TODO comment.

ziemleszcz commented on net/usbwlan.c:
```diff
@@ -0,0 +1,656 @@
+
+static int usbwlan_open(usbwlan_dev_t *dev, int flags, pid_t pid)
+{
+	if (dev->clientpid != 0 || dev->flags != 0) {
```
Common variables access should be protected by mutex. Otherwise this can fail for multiple message threads.

ziemleszcz commented on net/usbwlan.c:
```diff
@@ -0,0 +1,656 @@
+					if (msg.o.err >= 0) {
+						mutexUnlock(dev->rx.lock);
+						usbwlan_put(dev);
+						continue;
```
Notice that next `msgRecv`  call is also using `dev->rx.msg`.

julianuziemblo commented on net/usbwlan.c:
```diff
@@ -0,0 +1,656 @@
+
+		/* A device can be opened only by one process */
+		/* FIXME: On process exit, we get mtClose with pid=1 (as of 21.10.2025) */
+		if (((msg.type != mtOpen) && (msg.type != mtClose)) && (msg.pid != dev->clientpid && msg.pid != 1)) {
```
from the kernel side: no, because of refcounting there. The only exception is pid=1 (idle) which "cleans up" after a process exit, closing its fds - that's why this exception here

ziemleszcz commented on net/usbwlan.c:
```diff
@@ -0,0 +1,686 @@
+			case mtRead:
+				mutexLock(dev->rx.lock);
+				if (dev->rx.isPending) {
+					msg.o.err = -EINTR;
```
Why `EINTR`?

ziemleszcz commented on net/usbwlan.c:
```diff
@@ -0,0 +1,686 @@
+		}
+		else {
+			memcpy(dev->rx.msg.o.data, data, len);
+			dev->rx.msg.o.size = len;
```
Why set `size`?

ziemleszcz commented on net/usbwlan.c:
```diff
@@ -0,0 +1,685 @@
+	}
+
+	mutexLock(dev->rx.lock);
+	dev->rx.isPending = false;
```
Note that another thread may be waiting for `mtRead` response if `isPending` is true (this can happen when `USBWLAN_N_MSGTHREADS` > 1). You should consider sending the response here, or fix `USBWLAN_N_MSGTHREADS` to 1 (is there any real use case for multiple message threads?). Otherwise, the thread may end up waiting for `mtRead` response forever.

Darchiv reviewed (dismissed):
LGTM. Thanks for addressing the issues.

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,722 @@
+		usb_urbFree(dev->drv, dev->pipeBulkIn, dev->rx.urbid);
+	}
+	resourceDestroy(dev->rx.lock);
+	destroy_dev(dev->path);
```
Note that `dev->path` could be an empty string if `usbwlan_handleInsertion()` failed.

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,722 @@
+		msg->o.err = usb_transferAsync(dev->drv, dev->pipeBulkIn, dev->rx.urbid, USBWLAN_BULK_SIZE, NULL);
+		if (msg->o.err >= 0) {
+			dev->rx.isPending = true;
+			return EWOULDBLOCK;
```
-EWOULDBLOCK?

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,722 @@
+				mutexLock(dev->rx.lock);
+				err = _usbwlan_bulkReceive(dev, &msg, rid);
+				mutexUnlock(dev->rx.lock);
+				if (err == EWOULDBLOCK) {
```
-EWOULDBLOCK?

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,722 @@
+	}
+
+	if (dev->rx.isPending) {
+		msg->o.err = -EBUSY;
```
Maybe just return -EBUSY and let msg thread set `o.err`?

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,722 @@
+					usbwlan_put(dev);
+					continue;
+				}
+				break;
```
Set `msg.o.err = err` in else path?

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,722 @@
+#define USBWLAN_MSGTHREAD_STACK_SIZE 1024
+#endif
+
+#define USBWLAN_N_MSGTHREADS 1
```
If the code does not support more than one msg thread, please add a comment documenting this limitation.

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,719 @@
+
+	if (usbwlan_common.msgport != 0) {
+		msg_t msg = { 0 };
+		msg.type = mtDestroy;
```
`mtDestroy` handler was removed from `usbwlan_msgthread()`. How does this work?

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,719 @@
+		portDestroy(usbwlan_common.msgport);
+		usbwlan_common.msgport = 0;
+		if (usbwlan_common.msgtid != 0) {
+			threadJoin(usbwlan_common.msgtid, 0);
```
How does `usbwlan_msgthread()` end? By receiving error on `recvMsg()`?

julianuziemblo replied:
Yes - when port is destroyed, we receive error and exit

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,719 @@
+		usbwlan_common.lock = 0;
+	}
+
+	memset(&usbwlan_common, 0, sizeof(usbwlan_common));
```
Individual assignments that set `usb_common` fields to 0 are redundant and can be removed.
Was the stack clearing intentional, or is it a security feature?

julianuziemblo replied:
Wasn't the main focus, but while we're at it we can clear it I think - its only 1k. Its the error path anyway and device is closing so additional time spent here shouldnt be a problem IMO

ziemleszcz commented on net/usbwlan/usbwlan.c:
```diff
@@ -0,0 +1,716 @@
+	if (usbwlan_common.msgport != 0) {
+		msg_t msg = { 0 };
+		msg.type = mtDestroy;
+		(void)msgSend(usbwlan_common.msgport, &msg);
```
Is this message still required?

julianuziemblo replied:
rebase bug, sorry
